### ISSUE TITLE
update_app_identifier does not update project file unless Info.plist has identifier set to env var

### ISSUE THREAD

**Issue by crystalneth** [fastlane/fastlane#17014]
### Issue Description
update_app_identifier updates only Info.plist and not project file if Info.plist CFBundleIdentifier is set to interpolate from the environment, otherwise it updates only the project file.

Expected behavior seems to be to update both always, or at least to always update the project file. If nothing else this behavior should be documented.

Perhaps this was meant to skip update Info.plist if it was set to use the environment and was incorrectly implemented?

See line 17 & 18 here: https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/actions/update_app_identifier.rb
```
    app_id_equals_bundle_id = %W($(#{identifier_key}) ${#{identifier_key}}).include?(plist['CFBundleIdentifier'])
    if app_id_equals_bundle_id
         # update project file
    else
         # update Info.plist
    end
```

was maybe meant to be

```
    app_id_equals_bundle_id = %W($(#{identifier_key}) ${#{identifier_key}}).include?(plist['CFBundleIdentifier'])
    if !app_id_equals_bundle_id
         # update Info.plist
    end
     # update project file
```

##### Command executed
```
    update_app_identifier(
      xcodeproj: xcodeproj,
      plist_path: ENV['IOS_PLIST_PATH'],
      app_identifier: ENV['IOS_APP_ID']
    )
```

##### Fastlane env

<details><summary>✅ fastlane environment ✅</summary>
[](url)
### Stack

| Key                         | Value                                                          |
| --------------------------- | -------------------------------------------------------------- |
| OS                          | 10.15.6                                                        |
| Ruby                        | 2.6.3                                                          |
| Bundler?                    | false                                                          |
| Git                         | git version 2.22.0                                             |
| Installation Source         | /usr/local/bin/fastlane                                        |
| Host                        | Mac OS X 10.15.6 (19G73)                                       |
| Ruby Lib Dir                | /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib |
| OpenSSL Version             | LibreSSL 2.8.3                                                 |
| Is contained                | false                                                          |
| Is homebrew                 | false                                                          |
| Is installed via Fabric.app | false                                                          |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/                    |
| Xcode Version               | 11.6                                                           |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
fastlane_require 'dotenv'
fastlane_require 'spaceship'
release_notes_command = "git log HEAD --pretty=format:\"%s\" -1"

# Check Git Status

lane :check_git_status do |options|
  ensure_git_branch(
    branch: "#{options[:name]}"
  )
  ensure_git_status_clean
  git_pull

end

lane :set_build_numbers_to_current_timestamp do |options|
  incremented_build_number = Time.now.to_i.to_s
  `sed -i -e "s#.*IOS_VERSION_BUILD_NUMBER=.*#IOS_VERSION_BUILD_NUMBER='#{incremented_build_number}'#g" .env`
  `sed -i -e "s#.*ANDROID_VERSION_CODE=.*#ANDROID_VERSION_CODE='#{incremented_build_number}'#g" .env`
  ENV['IOS_VERSION_BUILD_NUMBER'] = incremented_build_number
  ENV['ANDROID_VERSION_CODE'] = incremented_build_number
end

# JS Environments

lane :set_js_env do |options|
  backup_file(path: 'src/environment/index.js')
  sh "cp ../src/environment/index.#{ENV['ENV']}.js ../src/environment/index.js"
end

before_all do |lane, options|
  Dotenv.load(".env.#{ENV['ENV']}.secret")
end

after_all do |lane, options|
  if File.exist?('../src/environment/index.js.back')
    restore_file(path: 'src/environment/index.js')
  end
end

error do |lane, exception, options|
  if File.exist?('../src/environment/index.js.back')
    restore_file(path: 'src/environment/index.js')
  end
end

# IOS

platform :ios do
  lane :restore_files do |options|
    ['config/GoogleService-Info.plist', '.env', "#{ENV['IOS_PROJECT_PATH']}/#{ENV['IOS_PLIST_PATH']}",
    "#{ENV['IOS_PROJECT_PATH']}/#{ENV['IOS_PROJECT_NAME']}.xcodeproj/project.pbxproj", "config/AppCenter-Config.plist"].each do |path|

      if File.exist?("../#{path}.back")
        restore_file(path: path)
      end
    end
    # appcenter_plist_path="#{ENV['IOS_PROJECT_PATH']}/#{ENV['IOS_PROJECT_NAME']}/AppCenter-Config.plist"
  end

  error do |lane, exception, options|
    restore_files
  end

  lane :setup do |options|
    produce(
      app_name: ENV['IOS_APP_NAME'],
      # skip_itc: true
    )
    match(
      type: 'development',
      shallow_clone: true,
      clone_branch_directly: true,
      force_for_new_devices: true
    )
    match(
      shallow_clone: true,
      clone_branch_directly: true,
      force_for_new_devices: true
    )
  end

  lane :setup_push do
    Spaceship.login
    Spaceship.select_team
    keyName = "Push JWT #{ENV['IOS_APP_NAME']} #{ENV['ENV']}".gsub!(/[^0-9A-Za-z ]/, '')
    key = Spaceship::Portal::Key.create(name: keyName, apns: true)
    puts "Key ID is"
    puts key.id
    File.write("#{keyName}.p8", key.download)
  end

  lane :update_files do
    google_service_info = "config/GoogleService-Info.plist"
    backup_file(path: google_service_info)
    rsync(source: "config/firebase/GoogleService-Info.#{ENV['ENV']}.plist", destination: google_service_info)
    plist_full_path = "#{ENV['IOS_PROJECT_PATH']}/#{ENV['IOS_PLIST_PATH']}"
    backup_file(path: plist_full_path)
    xcodeproj = "#{ENV['IOS_PROJECT_PATH']}/#{ENV['IOS_PROJECT_NAME']}.xcodeproj"
    pbxproj_full_path="#{xcodeproj}/project.pbxproj"
    backup_file(path: pbxproj_full_path)

    google_url_scheme = get_info_plist_value(path: google_service_info, key: 'REVERSED_CLIENT_ID')
    google_client_id = get_info_plist_value(path: google_service_info, key: 'CLIENT_ID')

    # backup_file(path: '.env')

    # sh "sed -ir -e \"s/GOOGLE_WEB_CLIENT_ID=.*/GOOGLE_WEB_CLIENT_ID=\\\"#{google_client_id}\\\"/\" ../.env"

    
    update_info_plist(
      xcodeproj: xcodeproj,
      plist_path: ENV['IOS_PLIST_PATH'],
      block: lambda { |plist|
        plist['CFBundleIdentifier'] = ENV['IOS_APP_ID']
        plist['CFBundleName'] = ENV['IOS_APP_NAME']
        plist['CFBundleDisplayName'] = ENV['IOS_APP_NAME']
        plist['CFBundleShortVersionString'] = ENV['IOS_VERSION']
        plist['CFBundleVersion'] = ENV['IOS_VERSION_BUILD_NUMBER']
        googleUrlScheme = plist["CFBundleURLTypes"].find{|scheme| scheme["CFBundleURLName"] == 'google' }
        googleUrlScheme['CFBundleURLSchemes'] = [google_url_scheme]
      },
    )
    if ENV['IOS_CODEPUSH_DEPLOYMENT_KEY'] then
      set_info_plist_value(path: plist_full_path, key:'CodePushDeploymentKey', value: ENV['IOS_CODEPUSH_DEPLOYMENT_KEY'])
    end
    if ENV['IOS_APPCENTER_APP_SECRET'] then
      # appcenter_plist_full_path = "#{ENV['IOS_PROJECT_PATH']}/#{ENV['IOS_PROJECT_NAME']}/AppCenter-Config.plist"
      appcenter_plist_full_path = "config/AppCenter-Config.plist"
      backup_file(path: appcenter_plist_full_path)
      set_info_plist_value(path: appcenter_plist_full_path, key:'AppSecret', value: ENV['IOS_APPCENTER_APP_SECRET'])
    end
    update_app_identifier(
      xcodeproj: xcodeproj,
      plist_path: ENV['IOS_PLIST_PATH'],
      app_identifier: ENV['IOS_APP_ID']
    )
  end

  lane :build do |options|
    update_files
    xcworkspace = "#{ENV['IOS_PROJECT_PATH']}/#{ENV['IOS_PROJECT_NAME']}.xcworkspace"
    profile_env_name = "sigh_#{ENV['IOS_APP_ID']}_#{ENV['MATCH_TYPE']}_profile-name"
    gymOptions = ({
      silent: true,
      xcargs: "PROVISIONING_PROFILE_SPECIFIER='#{ENV[profile_env_name]}' DEVELOPMENT_TEAM='#{ENV['IOS_TEAM_ID']}'"
    }).merge(
      File.directory?("../#{xcworkspace}") ?
        {workspace: xcworkspace} :
        {project: xcodeproj}
    )
    gym(gymOptions)
    restore_files
  end

  lane :deploy_hockey do |options|
    hockey(
      notify: '0',
      ipa: ENV['IOS_IPA_PATH']
    )
  end

  lane :deploy_to_appcenter do |options|
    appcenter_upload(
      api_token: ENV['FL_APPCENTER_API_TOKEN'],
      owner_name: ENV['APPCENTER_USERNAME'],
      app_name: ENV['IOS_APPCENTER_APP_ID'],
      ipa: ENV['IOS_IPA_PATH'],
      release_notes: %x[#{release_notes_command}]
    )
  end

  lane :deploy_to_testflight do |options|
    pilot(
      username: ENV['IOS_APPSTORECONNECT_USER_ID'],
      app_identifier: ENV['IOS_APP_ID'],
      ipa: ENV['IOS_IPA_PATH'],
      distribute_external: false,
      skip_waiting_for_build_processing: true
    )
  end

  lane :deploy do |options|
    set_js_env
    if options[:codepush] then
      release_notes = %x[#{release_notes_command}]
      sh "cd .. && appcenter codepush release-react -d #{ENV['IOS_CODEPUSH_DEPLOYMENT_NAME']} -a #{ENV['APPCENTER_USERNAME']}/#{ENV['IOS_APPCENTER_APP_ID']} --target-binary-version \"#{ENV['IOS_VERSION']}\" --description \"#{release_notes}\" --disable-duplicate-release-error"
    else
      setup_circle_ci
      match(
        shallow_clone: true,
        clone_branch_directly: true,
        readonly: true
      )
      build
      if ENV['DEPLOYMENT_PLATFORM'] === 'hockeyapp' then
        deploy_hockey
      elsif ENV['DEPLOYMENT_PLATFORM'] === 'appcenter' then
        deploy_to_appcenter
      elsif ENV['DEPLOYMENT_PLATFORM'] === 'appstore'
        deploy_to_testflight
      end
    end
  end

  lane :deploy_local do |options|
    check_git_status(name: ENV['REPO_GIT_BRANCH'])
    deploy
  end

  lane :add_devices do |options|
    if ['development','adhoc'].include?(ENV['MATCH_TYPE']) then
      register_devices(
        team_id: ENV['IOS_TEAM_ID'],
        devices_file: "./fastlane/ios_devices.txt"
      )
      match(
        shallow_clone: true,
        clone_branch_directly: true,
        force_for_new_devices: true
      )
    else
      puts 'Environment not associated to an ad-hoc or development app'
    end
  end

  lane :get_certificates_and_profiles do |options|
    match(
      type: 'development',
      shallow_clone: true,
      clone_branch_directly: true,
      readonly: true
    )
    match(
      shallow_clone: true,
      clone_branch_directly: true,
      readonly: true
    )
  end

end

#  ANDROID

platform :android do
  lane :restore_files do |options|
    appcenter_config_path = "#{ENV['ANDROID_PROJECT_DIR']}/app/src/main/assets/appcenter-config.json"
    if File.exist?("../#{appcenter_config_path}.back")
      restore_file(path: appcenter_config_path)
    end
    android_strings_path = "#{ENV['ANDROID_PROJECT_DIR']}/app/src/main/res/values/strings.xml"
    if File.exist?("../#{android_strings_path}.back")
      restore_file(path: android_strings_path)
    end
  end

  error do |lane, exception, options|
    restore_files
  end

  lane :set_keys do |options|
    if ENV['ANDROID_APPCENTER_APP_SECRET'] then
      appcenter_config_path = "#{ENV['ANDROID_PROJECT_DIR']}/app/src/main/assets/appcenter-config.json"
      backup_file(path: appcenter_config_path)
      appcenter_config = load_json(json_path: appcenter_config_path)
      appcenter_config['app_secret'] = ENV['ANDROID_APPCENTER_APP_SECRET']
      File.open("../#{appcenter_config_path}","w") do |f|
        f.write(appcenter_config.to_json)
      end
    end
    if ENV['ANDROID_CODEPUSH_DEPLOYMENT_KEY'] then
      android_strings_path = "#{ENV['ANDROID_PROJECT_DIR']}/app/src/main/res/values/strings.xml"
      # backup_file(path: android_strings_path) We can't backup this way because android can't build if there is a non-xml file in values folder. Solution to be found.
      xml = Nokogiri::XML(File.open("../#{android_strings_path}"))
      xml.at("//string[@name=\"reactNativeCodePush_androidDeploymentKey\"]").content = ENV['ANDROID_CODEPUSH_DEPLOYMENT_KEY']
      File.write("../#{android_strings_path}", xml.to_xml)
    end
  end

  lane :build do |options|
    set_keys
    gradle(
      task: "assembleRelease",
      project_dir: ENV['ANDROID_PROJECT_DIR']
    )
    restore_files
  end

  lane :deploy_hockey do |options|
    hockey(
      notify: '0',
      apk: ENV['ANDROID_APK_PATH']
    )
  end

  lane :deploy_to_appcenter do |options|
    appcenter_upload(
      api_token: ENV['FL_APPCENTER_API_TOKEN'],
      owner_name: ENV['APPCENTER_USERNAME'],
      app_name: ENV['ANDROID_APPCENTER_APP_ID'],
      apk: ENV['ANDROID_APK_PATH'],
      release_notes: %x[#{release_notes_command}]
    )
  end

  lane :deploy_to_playstore do |options|
    supply(
      package_name: ENV['GRADLE_APP_IDENTIFIER'],
      track: 'internal',
      apk: ENV['ANDROID_APK_PATH'],
      json_key: ENV['ANDROID_PLAYSTORE_JSON_KEY_PATH'],
    )
  end

  lane :deploy do |options|
    set_js_env
    if options[:codepush] then
      release_notes = %x[#{release_notes_command}]
      sh "cd .. && appcenter codepush release-react -d #{ENV['ANDROID_CODEPUSH_DEPLOYMENT_NAME']} -a #{ENV['APPCENTER_USERNAME']}/#{ENV['ANDROID_APPCENTER_APP_ID']} --target-binary-version \"#{ENV['ANDROID_VERSION_NAME']}\" --description \"#{release_notes}\" --disable-duplicate-release-error"
    else
      build
      if ENV['DEPLOYMENT_PLATFORM'] === 'hockeyapp' then
        deploy_hockey
      elsif ENV['DEPLOYMENT_PLATFORM'] === 'appcenter' then
        deploy_to_appcenter
      elsif ENV['DEPLOYMENT_PLATFORM'] === 'appstore' then
        deploy_to_playstore
      end
    end
  end

  lane :deploy_local do |options|
    check_git_status(name: ENV['REPO_GIT_BRANCH'])
    deploy
  end

end

```
</details>

<details><summary>`./fastlane/Appfile`</summary>

```ruby
# See https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Appfile.md
app_identifier ENV['IOS_APP_ID']
apple_id ENV['IOS_USER_ID']
team_id ENV['IOS_TEAM_ID']
itc_team_name ENV['IOS_ITC_TEAM_NAME']

```
</details>

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.155.1 | ✅ Up-To-Date  |

### Loaded fastlane plugins:

| Plugin                    | Version | Update-Status |
| ------------------------- | ------- | ------------- |
| fastlane-plugin-badge     | 1.4.0   | ✅ Up-To-Date  |
| fastlane-plugin-appcenter | 1.9.0   | ✅ Up-To-Date  |
| fastlane-plugin-load_json | 0.0.1   | ✅ Up-To-Date  |

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.4.0        |
| slack-notifier            | 2.3.2        |
| rouge                     | 2.0.7        |
| xcpretty                  | 0.3.0        |
| terminal-notifier         | 2.0.0        |
| unicode-display_width     | 1.7.0        |
| terminal-table            | 1.8.0        |
| plist                     | 3.5.0        |
| addressable               | 2.7.0        |
| multipart-post            | 2.0.0        |
| word_wrap                 | 1.0.0        |
| tty-cursor                | 0.7.1        |
| tty-spinner               | 0.9.3        |
| babosa                    | 1.0.3        |
| colored                   | 1.2          |
| highline                  | 1.7.10       |
| commander-fastlane        | 4.4.6        |
| faraday                   | 1.0.1        |
| unf_ext                   | 0.0.7.7      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20190701 |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| faraday_middleware        | 1.0.0        |
| gh_inspector              | 1.1.3        |
| mini_magick               | 4.10.1       |
| rubyzip                   | 2.3.0        |
| security                  | 0.1.3        |
| xcpretty-travis-formatter | 1.0.0        |
| bundler                   | 2.1.4        |
| naturally                 | 2.2.0        |
| simctl                    | 1.6.8        |
| uber                      | 0.1.0        |
| declarative-option        | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 3.1.2        |
| mini_mime                 | 1.0.2        |
| signet                    | 0.14.0       |
| httpclient                | 2.8.3        |
| google-api-client         | 0.38.0       |
| nanaimo                   | 0.3.0        |
| colored2                  | 3.1.2        |
| claide                    | 1.0.3        |
| CFPropertyList            | 3.0.2        |
| atomos                    | 0.1.3        |
| xcodeproj                 | 1.17.1       |
| public_suffix             | 4.0.5        |
| tty-screen                | 0.8.1        |
| excon                     | 0.76.0       |
| fastimage                 | 2.2.0        |
| json                      | 2.3.1        |
| dotenv                    | 2.7.6        |
| jwt                       | 2.2.1        |
| declarative               | 0.0.20       |
| multi_json                | 1.15.0       |
| os                        | 1.1.1        |
| memoist                   | 0.16.2       |
| googleauth                | 0.13.1       |
| rake                      | 13.0.1       |
| digest-crc                | 0.6.1        |
| google-cloud-errors       | 1.0.1        |
| google-cloud-env          | 1.3.3        |
| google-cloud-core         | 1.5.0        |
| google-cloud-storage      | 1.27.0       |
| emoji_regex               | 3.0.0        |
| aws-eventstream           | 1.1.0        |
| aws-sigv4                 | 1.2.1        |
| aws-partitions            | 1.352.0      |
| jmespath                  | 1.4.0        |
| aws-sdk-core              | 3.104.3      |
| aws-sdk-kms               | 1.36.0       |
| aws-sdk-s3                | 1.75.0       |
| forwardable               | 1.3.1        |
| logger                    | 1.4.2        |
| stringio                  | 0.0.2        |
| ipaddr                    | 1.2.2        |
| openssl                   | 2.1.2        |
| ostruct                   | 0.1.0        |
| strscan                   | 1.0.0        |
| date                      | 3.0.0        |
| fileutils                 | 1.4.1        |
| io-console                | 0.5.6        |
| zlib                      | 1.0.0        |
| libxml-ruby               | 3.2.0        |
| rexml                     | 3.1.9        |
| psych                     | 3.1.0        |
| mutex_m                   | 0.1.0        |
| webrick                   | 1.4.2        |
| curb                      | 0.9.10       |
| badge                     | 0.12.0       |
| fastlane-plugin-badge     | 1.4.0        |
| fastlane-plugin-appcenter | 1.9.0        |
| csv                       | 3.1.5        |
| fastlane-plugin-load_json | 0.0.1        |
</details>

*generated on:* **2020-08-06**
</details>

**Comment by crystalneth** [comment]
I've added that

**Comment by crystalneth** [comment]
https://github.com/fastlane/fastlane/issues/15584 and https://github.com/fastlane/fastlane/issues/15500 look the same, but they didn't get addressed.

**Comment by iBotPeaches** [comment]
Adding a comment as I believe this is what I've tracked down.

My blob

```
    update_app_identifier(
      app_identifier: "net.redacted.dev",
      xcodeproj: "App.xcodeproj",
      plist_path: "App/Info.plist"
    )
```

The error
```
[18:20:38]: ▸     User-supplied CFBundleIdentifier value 'net.redacted.dev' in the Info.plist must be the same as the PRODUCT_BUNDLE_IDENTIFIER build setting value 'net.redacted.normal'. (in target 'App' from project 'App')
```

As digging into it, it is exactly as mentioned above. I started looking at the code and it looks like it is inverted accidentally, since if `app_id_equals_bundle_id` it seems like you wouldn't want to edit the xcodeproj. If they were different (like they are for me) then we do need to switch it out.

I made that switch locally and my local worked perfectly, but then every test for that action failed as expected. I'm just trying to figure out if I'm not thinking this through right.

I assume many try to rename the bundle at build time for beta/test apps, but this xcodeproj not being auto renamed is preventing that.
